### PR TITLE
Fixed N+1 queries issue in /api/v1/bed/

### PR DESCRIPTION
**Pull request by dhruv-goyal-10**
## Proposed Changes

- **is_occupied** is the property defined in the **Bed Model**, which was contributing to N + 1 queries issue, and I have annotated the attribute **_is_occupied** in the queryset, which basically corresponds to the same value as the **is_occupied property** in Bed Model. I have handled the various attribute cases, in **BedSerializer**, by defining **is_occupied** field as **SerializerMethodField**, hence there will be no change in API response.

- Have added **location__facility** in the select_related field, as this was also contributing to the N + 1 queries issue.

### Associated Issue

- Solves #1338

## Merge Checklist

- [x] Tests added/fixed
- [ ] Update docs in `/docs`
- [x] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

**Comment by dhruv-goyal-10** [comment]
@sainak @rithviknishad @vigneshhari Can you please review this PR?

**Comment by sainak** [comment]
hey @dhruv-goyal-10 thanks for the pr,
after looking at the bed model i can see that the is_occupied property on the model is not being used
https://github.com/coronasafe/care/blob/6cb4b3f26c03bece09f818299f4dccfb321ab86b/care/facility/models/bed.py#L43-L45

so we can just annotate the queryset with is_occupied directly if `self.action == list`

**Comment by dhruv-goyal-10** [comment]
@sainak @rithviknishad @vigneshhari I have completed the changes as suggested by @sainak. Can you review my PR now?

File: care/facility/api/serializers/bed.py
Comment (sainak) [inline care/facility/api/serializers/bed.py:1]: is_occupied = BooleanField(default=False, read_only=True)
is correct, you can revert the changes in this file

**Comment by dhruv-goyal-10** [comment]
Hey @sainak! I have done changes suggested by you. Can you please have a look at my PR now?

**Review by sainak — APPROVED** [review]

**Review by Ashesh3 — APPROVED** [review]

**Comment by sainak** [comment]
> @sainak I later realised this, commenting out the is_occupied property will although not affect the results in case of list method, but it will affect the result in case of retrieve method (fetching a particular object using id), coz we aren't annotating is_occupied property in that case and it will always corresponds to False (the default value).
> Isn't it?

Currently we are not using bed retrieve on Fe but you are correct that will give inconsistent results, It won't affect the performance so lets remove the if condition

File: care/facility/api/viewsets/bed.py [hunk 1]
```diff
@@ -63,6 +63,16 @@ class BedViewSet(
...
         queryset = self.queryset
+
+        if self.action == "list":
+            queryset = queryset.annotate(
+                is_occupied=Exists(
+                    ConsultationBed.objects.filter(
+                        bed__id=OuterRef("id"), end_date__isnull=True
+                    )
+                )
+            )
```
Comment (sainak) [inline care/facility/api/viewsets/bed.py:74]:
```suggestion
        queryset = queryset.annotate(
            is_occupied=Exists(
                ConsultationBed.objects.filter(
                    bed__id=OuterRef("id"), end_date__isnull=True
                )
            )
        )
```

**Comment by dhruv-goyal-10** [comment]
@sainak Made the changes as suggested by you.
Also, thanks for an insightful discussion.

**Review by rithviknishad — APPROVED** [review]

**Comment by dhruv-goyal-10** [comment]
@sainak Do I need to write test cases for this PR to get merged?

**Comment by sainak** [comment]
Yes!
For this case create a bunch of records and checked the number of queries executed

**Comment by dhruv-goyal-10** [comment]
@sainak I have added the test case for this as suggested by you, Please have a look now!

File: care/facility/tests/test_bed_api.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        )
+        self.bed2 = Bed.objects.create(
+            name="bed2", location=self.asset_location, facility=self.facility
+        )
+        self.bed3 = Bed.objects.create(
+            name="bed3", location=self.asset_location, facility=self.facility
+        )
+
+    def test_list_beds(self):
+        with self.assertNumQueries(5):
```
Comment (sainak) [inline care/facility/tests/test_bed_api.py:35]:
```suggestion
        # includes 3 queries for auth and 1 for pagination count
        with self.assertNumQueries(5):
```

**Review by sainak — APPROVED** [review]

**Review by Ashesh3 — APPROVED** [review]

**Review by vigneshhari — APPROVED** [review]
